ably/realtime_conn.go

```diff
@@ -86,23 +90,78 @@ func newConn(opts *clientOptions, auth *Auth, callbacks connCallbacks) (*Connect
 		callbacks: callbacks,
 	}
 	c.queue = newMsgQueue(c)
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.watchDisconnect(c.ctx)
 	if !opts.NoConnect {
-		if _, err := c.connect(false); err != nil {
+		if _, err := c.connect(connArgs{}); err != nil {
 			return nil, err
 		}
 	}
 	return c, nil
 }
 
-func (c *Connection) dial(proto string, u *url.URL) (proto.Conn, error) {
+func (c *Connection) dial(proto string, u *url.URL, once bool) (conn proto.Conn, err error) {
+	lg := c.logger().Sugar()
+	start := time.Now()
+	lg.Debugf("Dial protocol=%q url %q ", proto, u.String())
 	// (RTN23b)
 	query := u.Query()
 	query.Add("heartbeats", "true")
 	u.RawQuery = query.Encode()
+	// We first try a single dial to see if we are successful. There is no need to
+	// have this only in the for loop because most cases we will succeed after the
+	// first attempt, let's be prudent here.
+	conn, err = c.dialInternal(proto, u, c.opts.realtimeRequestTimeout())
+	if err != nil {
+		lg.Debugf("Dial Failed in %v with %v", time.Since(start), err)
+		return nil, err
+	}
+	lg.Debugf("Dial success in %v", time.Since(start))
+	return conn, err
+}
+
+func (c *Connection) connectAfterSuspension(arg connArgs) (Result, error) {
+	timeout := c.opts.suspendedRetryTimeout()
+	lg := c.logger().Sugar()
+	lg.Debugf("Attemting to periodically establish connection after suspension with timeout %v", timeout)
+	tick := time.NewTicker(timeout)
+	for {
+		select {
+		case <-c.ctx.Done():
+			lg.Debug("exiting connetAfterSuspension loop")
+			return nil, c.ctx.Err()
+		case <-tick.C:
+			res, err := c.connectWithInternal(arg)
+			if err != nil {
+				if recoverable(err) {
+					continue
+				}
+				return nil, err
+			}
+			return res, nil
+		}
+	}
+}
+
+func (c *Connection) dialInternal(protocol string, u *url.URL, timeout time.Duration) (proto.Conn, error) {
 	if c.opts.Dial != nil {
-		return c.opts.Dial(proto, u)
+		return c.opts.Dial(protocol, u, timeout)
+	}
+	return ablyutil.DialWebsocket(protocol, u, timeout)
+}
+
+// recoverable returns true if err is recoverable, err is from making a
+// connection
+func recoverable(err error) bool {
```

Comment:
Need to define what a recoverable error must be.

Reply:
@gernest This function tries to exhaustively check all errors that are recoverable, and defaults to false otherwise. I'd say it needs to do the opposite: exhaustively check the handful of errors we know are unrecoverable (that is, 4xx response status from Ably, except token errors in the [40140, 40150) range for which there is some further logic) and default to true (recoverable) otherwise.

There can be all kinds of transient error conditions that aren't timeouts. Those shouldn't kill our retry loop.

Reply:
done at 80f6c2a and 8058306